library/std/src/sys/common/thread_local/fast_local.rs

```diff
@@ -74,170 +95,205 @@ pub macro thread_local_inner {
 }
 
-// This data structure has been carefully constructed so that the fast path
-// only contains one branch on x86. That optimization is necessary to avoid
-// duplicated tls lookups on OSX.
-//
-// LLVM issue: https://bugs.llvm.org/show_bug.cgi?id=41722
-pub struct Key<T> {
-    // If `LazyKeyInner::get` returns `None`, that indicates either:
-    //   * The value has never been initialized
-    //   * The value is being recursively initialized
-    //   * The value has already been destroyed or is being destroyed
-    // To determine which kind of `None`, check `dtor_state`.
-    //
-    // This is very optimizer friendly for the fast path - initialized but
-    // not yet dropped.
-    inner: LazyKeyInner<T>,
-
-    // Metadata to keep track of the state of the destructor. Remember that
-    // this variable is thread-local, not global.
-    dtor_state: Cell<DtorState>,
+pub trait DestroyedState {
+    const EXISTS: bool;
+    fn create() -> Self;
 }
 
-impl<T> fmt::Debug for Key<T> {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Key").finish_non_exhaustive()
+impl DestroyedState for ! {
+    const EXISTS: bool = false;
+    fn create() -> ! {
+        panic!("attempted to create nonexistent state");
     }
 }
-impl<T> Key<T> {
-    pub const fn new() -> Key<T> {
-        Key { inner: LazyKeyInner::new(), dtor_state: Cell::new(DtorState::Unregistered) }
-    }
 
-    // note that this is just a publicly-callable function only for the
-    // const-initialized form of thread locals, basically a way to call the
-    // free `register_dtor` function defined elsewhere in std.
-    pub unsafe fn register_dtor(a: *mut u8, dtor: unsafe extern "C" fn(*mut u8)) {
-        unsafe {
-            register_dtor(a, dtor);
-        }
-    }
+impl DestroyedState for () {
+    const EXISTS: bool = true;
+    #[inline]
+    fn create() {}
+}
 
-    pub unsafe fn get<F: FnOnce() -> T>(&self, init: F) -> Option<&'static T> {
-        // SAFETY: See the definitions of `LazyKeyInner::get` and
-        // `try_initialize` for more information.
-        //
-        // The caller must ensure no mutable references are ever active to
-        // the inner cell or the inner T when this is called.
-        // The `try_initialize` is dependant on the passed `init` function
-        // for this.
-        unsafe {
-            match self.inner.get() {
-                Some(val) => Some(val),
-                None => self.try_initialize(init),
-            }
-        }
+enum State<I, T, D> {
+    Initial(I),
+    Alive(T),
+    Destroyed(D),
+}
+
+pub struct Storage<I, T, D> {
+    state: UnsafeCell<State<I, T, D>>,
+}
+
+impl<I, T, D> Storage<I, T, D>
+where
+    D: DestroyedState,
+{
+    /// Create a new TLS storage with the provided initialization parameter.
+    pub const fn new(param: I) -> Storage<I, T, D> {
+        Storage { state: UnsafeCell::new(State::Initial(param)) }
     }
```

Comment:
There could be another function that initializes the storage directly to the alive state, that way the `const` can skip the initializer on the first call and eliminate the `Initial` variant.
For example, const could have the type `Storage<!, $t, ()>` (if `$t` needs drop). This could layout optimize down to the same layout as `$t` if it has exactly one niche instead of `Storage<$t, $t, ()>`, which can't be layout optimized.